src/details/ArborX_MinimumSpanningTree.hpp

```diff
@@ -0,0 +1,504 @@
+                      Radii const &radii)
+{
+  auto const n = bvh.size();
+  Kokkos::parallel_for(
+      "ArborX::MST::reset_shared_radii",
+      Kokkos::RangePolicy<ExecutionSpace>(space, n - 1, 2 * n - 2),
+      KOKKOS_LAMBDA(int i) {
+        int const j = i + 1;
+        auto const label_i = labels(i);
+        auto const label_j = labels(j);
+        if (label_i != label_j)
+        {
+          auto const r = mutual_reachability_distance(
+              HappyTreeFriends::getLeafPermutationIndex(bvh, i),
+              HappyTreeFriends::getLeafPermutationIndex(bvh, j),
+              distance(HappyTreeFriends::getBoundingVolume(bvh, i),
+                       HappyTreeFriends::getBoundingVolume(bvh, j)));
+          Kokkos::atomic_min(&radii(label_i - n + 1), r);
+          Kokkos::atomic_min(&radii(label_j - n + 1), r);
+        }
+      });
+}
+
+template <class MemorySpace>
+struct MinimumSpanningTree
+{
+  Kokkos::View<WeightedEdge *, MemorySpace> edges;
+
+  template <class ExecutionSpace, class BVH, class Metric>
+  void doBoruvka(ExecutionSpace const &space, BVH const &bvh,
+                 Metric const &mutual_reachability_distance)
+  {
+    auto const n = bvh.size();
+    Kokkos::View<int *, MemorySpace> parents(
+        Kokkos::view_alloc(Kokkos::WithoutInitializing, "ArborX::MST::parents"),
+        2 * n - 1);
+    findParents(space, bvh, parents);
+
+    Kokkos::Profiling::pushRegion("ArborX::MST::initialize_node_labels");
+    Kokkos::View<int *, MemorySpace> labels(
+        Kokkos::view_alloc(Kokkos::WithoutInitializing, "ArborX::MST::labels"),
+        2 * n - 1);
+    iota(space, Kokkos::subview(labels, std::make_pair(n - 1, 2 * n - 1)),
+         n - 1);
+    Kokkos::Profiling::popRegion();
+
+    Kokkos::View<WeightedEdge *, MemorySpace> component_out_edges(
+        Kokkos::view_alloc(Kokkos::WithoutInitializing,
+                           "ArborX::MST::component_out_edges"),
+        n);
+    constexpr auto inf = KokkosExt::ArithmeticTraits::infinity<float>::value;
+    Kokkos::deep_copy(space, component_out_edges, WeightedEdge{-1, -1, inf});
+
+    Kokkos::View<int, MemorySpace> num_edges(Kokkos::view_alloc(
+        Kokkos::WithoutInitializing, "ArborX::MST::num_edges"));
+    Kokkos::deep_copy(space, num_edges, 0);
+
+    Kokkos::View<float *, MemorySpace> radii(
+        Kokkos::view_alloc(Kokkos::WithoutInitializing, "ArborX::MST::radii"),
+        n);
+
+    // Boruvka iterations
+    int iterations = 0;
+    int num_components = n;
+    do
+    {
+      Kokkos::Profiling::pushRegion("ArborX::Boruvka_" +
+                                    std::to_string(++iterations) + "_" +
+                                    std::to_string(num_components));
+      reduceLabels(space, parents, labels);
+      Kokkos::deep_copy(space, component_out_edges, {-1, -1, inf});
+      Kokkos::deep_copy(space, radii, inf);
+      resetSharedRadii(space, bvh, labels, mutual_reachability_distance, radii);
```

Comment:
As with the previous comments, I think these lines should be moved inside the class.